crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4374,8 +4409,8 @@ mod tests {
         )?;
 
         assert_public_ty(&db, "src/a.py", "x", "Literal[3, 4, 5]");
-        assert_public_ty(&db, "src/a.py", "r", "Unbound | Literal[2]");
-        assert_public_ty(&db, "src/a.py", "s", "Unbound | Literal[5]");
+        assert_public_ty(&db, "src/a.py", "r", "Literal[2]");
+        assert_public_ty(&db, "src/a.py", "s", "Literal[5]");
```

Comment:
This is one of those cases where part of the whole purpose of the test is actually to show that we know which names might be unbound, so we should really update the test such that it doesn't use (only)`assert_public_ty`, because it's really not the public type we want to assert on.

I earlier suggested an approach using `reveal_type`, and I think that's actually a fine option in the tests below where you use it.

I realized a simpler approach in some of these tests (particularly the global-scope ones) might be to keep these `assert_public_ty` but add another helper `assert_may_be_unbound` (which uses the use-def map directly) and explicitly assert here that `r` and `s` may be unbound. This way we actually get both the assertion on the public type, and an assertion that locally we do know the name might be unbound.

Reply:
Is there a way to add public type tests to the markdown tests introduced in #13719?

Reply:
You can import it (since tests can be multi-file), or refer to the name as a global from within a nested function. Import probably makes it clearer what's going on?

Reply:
How do I run the mdtests? `cargo test` inside `red_knot_python_semantic` and `red_knot_test` both don't seem to run it.

Reply:
`cargo test` inside `red_knot_python_semantic` (or `cargo test -p red_knot_python_semantic` from the workspace) should run the mdtests, and do for me. How are you concluding that they aren't running?

Reply:
Hmm, they only run when I run `cargo test mdtest` inside `red_knot_python_semantic`.

Without `mdtest`: https://gist.github.com/pilleye/7454e668bc19f9ca1969e0765a82a301
With `mdtest`: https://gist.github.com/pilleye/3e9842e37a911eb88f6ac26c281bb0dc

Reply:
That's weird -- I'm not enough of a `cargo test` expert to explain this discrepancy :) For me they definitely run with just `cargo test`... In general my experience has been that `cargo test` runs all tests it discovers, both unit tests and integration tests. Not sure what might cause that to not be happening for you. `mdtest` is just a normal Rust integration test.

Reply:
huh, that _is_ counterintuitive!

Reply:
Ah, that makes sense! TIL